File: source_formatting/commentary_prefixes.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from source_formatting.html_parser import BaseHtmlTranslator

_PREFIXES_TO_STRIP = (
    "גמ'",
    "מתני'",
)

class CommentaryPrefixStripper(BaseHtmlTranslator):
    def __init__(self):
        super().__init__()
        self.has_processed_text = False

    def handle_starttag(self, tag, attrs):
        self.append_start_tag(tag, attrs)

    def handle_endtag(self, tag):
        self.append_end_tag(tag)

    def handle_data(self, data):
        if not self.has_processed_text:
            for prefix in _PREFIXES_TO_STRIP:
                prefix = f"{prefix} "
                if data.startswith(prefix):
                    data = data[len(prefix):]
            self.has_processed_text = True
        self._out.append(data)

File: source_formatting/commentary_prefixes.ts
import {Attributes, HtmlVisitor} from "./html_visitor";

const _PREFIXES_TO_STRIP = [
  "גמ'",
  "מתני'",
].flatMap(
  x => [x, x.replace("'", "׳")],
).map(x => `${x} `);

export class CommentaryPrefixStripper extends HtmlVisitor {
  hasProcessedText = false

  shouldRun(input: string): boolean {
    return _PREFIXES_TO_STRIP.some(x => input.includes(x));
  }

  visitStartTag(tag: string, attributes: Attributes): void {
    this.appendStartTag(tag, attributes);
  }

  visitEndTag(tag: string): void {
    this.appendEndTag(tag);
  }

  visitText(text: string): void {
    if (!this.hasProcessedText) {
      for (const prefix of _PREFIXES_TO_STRIP) {
        if (text.startsWith(prefix)) {
          text = text.substring(prefix.length);
        }
      }
      this.hasProcessedText = true;
    }

    this.appendText(text);
  }
}
